lib/lte_link_control/lte_lc.c

```diff
@@ -151,32 +182,328 @@ static const char cgauth[] = "AT+CGAUTH="CONFIG_LTE_PDN_AUTH;
+		goto clean_exit;
+	}
+
+	/* Parse network registration status */
+	err = at_params_int_get(&resp_list,
+				AT_CEREG_REG_STATUS_INDEX,
+				&status);
+	if (err) {
+		LOG_ERR("Could not get registration status, error: %d", err);
+		goto clean_exit;
+	}
+
+	*reg_status = status;
+
+	/* Parse tracking area code */
+	err = at_params_string_get(&resp_list,
+				   AT_CEREG_TAC_INDEX,
+				   str_buf, &len);
+	if (err) {
+		LOG_ERR("Could not get tracking area code, error: %d", err);
+		goto clean_exit;
+	}
+
+	str_buf[len] = '\0';
+	cell->tac = strtoul(str_buf, NULL, 16);
+
+	/* Parse cell ID */
+	len = sizeof(str_buf) - 1;
+
+	err = at_params_string_get(&resp_list,
+				   AT_CEREG_CELL_ID_INDEX,
+				   str_buf, &len);
+	if (err) {
+		LOG_ERR("Could not get cell ID, error: %d", err);
+		goto clean_exit;
+	}
+
+	str_buf[len] = '\0';
+	cell->id = strtoul(str_buf, NULL, 16);
+
+	/* Parse PSM configuration */
+	err = parse_psm_cfg(&resp_list, true, psm_cfg);
+	if (err) {
+		LOG_ERR("Failed to parse PSM configuration, error: %d", err);
+		goto clean_exit;
+	}
+
+clean_exit:
+	at_params_list_free(&resp_list);
+
+	return err;
+}
+
+static void at_handler(void *context, const char *response)
 {
 	ARG_UNUSED(context);
 
 	int err;
-	enum lte_lc_nw_reg_status status;
+	bool notify = false;
+	enum lte_lc_notif_type notif_type;
+	struct lte_lc_evt evt;
 
 	if (response == NULL) {
 		LOG_ERR("Response buffer is NULL-pointer");
 		return;
 	}
 
-	err = parse_nw_reg_status(response, &status, AT_CEREG_REG_STATUS_INDEX);
-	if (err) {
-		LOG_ERR("Could not get network registration status");
+	/* Only proceed with parsing if notification is relevant */
+	if (!is_relevant_notif(response, &notif_type)) {
 		return;
 	}
 
-	if ((status == LTE_LC_NW_REG_REGISTERED_HOME) ||
-	    (status == LTE_LC_NW_REG_REGISTERED_ROAMING)) {
-		k_sem_give(&link);
+	switch (notif_type) {
+	case LTE_LC_NOTIF_CEREG: {
+		static enum lte_lc_nw_reg_status prev_reg_status;
+		static struct lte_lc_cell prev_cell;
+		static struct lte_lc_psm_cfg prev_psm_cfg;
+		enum lte_lc_nw_reg_status reg_status = 0;
+		struct lte_lc_cell cell;
+		struct lte_lc_psm_cfg psm_cfg;
+
+		LOG_DBG("+CEREG notification: %s", log_strdup(response));
+
+		err = parse_cereg(response, &reg_status, &cell, &psm_cfg);
+		if (err) {
+			LOG_ERR("Notification could not be parsed, error: %d",
+				err);
+			return;
+		}
+
+		if ((reg_status == LTE_LC_NW_REG_REGISTERED_HOME) ||
+		    (reg_status == LTE_LC_NW_REG_REGISTERED_ROAMING)) {
+			k_sem_give(&link);
+		}
+
+		if (!evt_handler) {
+			return;
+		}
+
+		/* Network registration status event */
+		if (reg_status != prev_reg_status) {
+			prev_reg_status = reg_status;
+			evt.type = LTE_LC_EVT_NW_REG_STATUS;
+			evt.nw_reg_status = reg_status;
+
+			evt_handler(&evt);
+		}
+
+		/* Cell update event */
+		if (memcmp(&cell, &prev_cell, sizeof(struct lte_lc_cell))) {
+			evt.type = LTE_LC_EVT_CELL_UPDATE;
+
+			memcpy(&prev_cell, &cell, sizeof(struct lte_lc_cell));
```

Comment:
It should be enough to save the last event that was sent. Why are you saving two events?

Reply:
I'm copying the parsed cell information into the event that's about to be dispatched (and subsequently overwritten), and into `prev_cell` for comparison next time around

Reply:
`struct lte_lc_evt evt;` is allocated on the stack, `static struct lte_lc_cell prev_cell;` is not.
My bad, this looks okay.